File: trees/c/binary_tree.cpp
#include<iostream>

template<typename T>
class BinaryTree {
    public:
        T value;
        BinaryTree<T>* left;
        BinaryTree<T>* right;
 
    BinaryTree<T>(T value) {
        this->value = value;
        left = NULL;
        right = NULL;
    }
 
    void add(T value) {
        if(left && value < left->value) {
            left->add(value);
        }
        else if(right && value > right->value) {
            right->add(value);
        }
        else {
            if(this->value < value) {
                if(left) left->add(this->value);
                else     left = new BinaryTree<T>(this->value);
            }
            else {
                if(right) right->add(this->value);
                else      right = new BinaryTree<T>(this->value);
            }
            this->value = value;
        }
    }
    
    bool find(T value, BinaryTree<T>* a) {
        if(value > this->value && right) {
            return right->find(value, a);
        }
        else if(value < this->value && left) {
            return left->find(value, a);
        }
        else {
            if(value == this->value) {
                *a = *this;
                return true;
            }
            else {
                return false;
            }
        }
    }
 
    void display() {
        display(0);
    }
 
    void display(int depth) {
        display_offset(depth);
        std::cout << value << "\n";
        if(right) {
            right->display(depth + 1);
        }
        if(left) {
            left->display(depth + 1);
        }
    }
 
 
    private: 
 
    void display_offset(int depth) {
        for(int i = 0; i < depth; i++) {
            std::cout << ' ';
        }
    }
};
 
int main() {
    BinaryTree<float> b = BinaryTree<float>(2.2);
    b.display();
    b.add(2.4);
    b.display();
    
    b.add(4.8);
    b.display();
    b.add(-2.2);
    b.display();
    BinaryTree<float>* f = new BinaryTree<float>(0.0);
    if(b.find(2.4, f)){f->display();}
    else{std::cout << "NULL\n";}
}

File: trees/python/binary_tree.py
class BinaryTree:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None

    def add(self, value):
    	if self.left and value < self.left.value:
    		self.left.add(value)
    	elif self.right and value > self.right.value:
    		self.right.add(value)
    	else:
    		if self.value < value:
    			if self.left:
    				self.left.add(self.value)
    			else:
    				self.left = BinaryTree(self.value)
    		else:
    			if self.right:
    				self.right.add(self.value)
    			else:
    				self.right = BinaryTree(self.value)
    		self.value = value

    def find(self, value):
    	if value > self.value and self.right:
    		return self.right.find(value)
    	elif value < self.value and self.left:
    		return self.left.find(value)
    	else:
    		if value == self.value:
    			return self
    		else:
    			return None

    def __str__(self):
    	return self.as_string(0)

    def as_string(self, depth):
    	return "%s%s\n%s%s\n" % (
    		' ' * depth,
    		self.value,
    		self.right.as_string(depth + 1) if self.right else '',
    		self.left.as_string(depth + 1) if self.left else '')
    
    __repr__ = __str__

b = BinaryTree(2.2)
print(b)
b.add(-2)
print(b)

print(b.find(-2))

